lib/wikidata.rb

```diff
@@ -18,17 +18,50 @@ def self.ids_from_pages(lang, titles)
         titles:     sliced.join('|'),
         token_type: false,
       }
-      response = client.action :query, page_args
-      redirected_from = Hash[(response.data['redirects'] || []).map { |h| [h['to'], h['from']] }]
-      response.data['pages'].select { |_k, v| v.key? 'pageprops' }.map do |_k, v|
-        [redirected_from[v['title']] || v['title'], v['pageprops']['wikibase_item']]
-      end
+      TitlesAndIds.new(titles, client.action(:query, page_args)).all
     end
     results = Hash[res.flatten(1)]
     missing = titles - results.keys
     warn "Can't find Wikidata IDs for: #{missing.join(', ')} in #{lang}" if missing.any?
     results
   end
+
+  class TitlesAndIds
+    def initialize(titles, response)
+      @response = response
+      @titles = titles
+    end
+
+    def all
+      # response.data['pages'] contains the direct title of the article
+      # regardless of whether the query contains the direct title or not. To
+      # ensure we return only the titles requested, we map the results to the
+      # titles provided, so that any unrequested direct titles are not
+      # included in the result.
+      titles.reduce([]) do |arr, title|
+        arr << [title, titles_hash[title]] unless titles_hash[title].nil?
```

Comment:
`reduce` is almost always going to be the wrong approach to this sort of thing. 

I'm not entirely sure I understand what's going on here, but this seems to be simpler and clearer way of doing the same thing

```ruby
titles.reject { |t| titles_hash[t].nil? }.map {|t| [t, titles_hash[t]] }
```